### ISSUE TITLE
ale_lint_on_* are not working

### ISSUE THREAD

**Issue by thornycrackers** [w0rp/ale#1294]
## Information

NVIM v0.2.2
Build type: RelWithDebInfo
LuaJIT 2.0.4

Ubuntu 16.04

### :ALEInfo

PASTE OUTPUT OF `:ALEInfo` HERE. YOU CAN TRY `:ALEInfoToClipboard`.

Enabled Linters: ['flake8', 'mypy', 'pylint']
 Linter Variables:
let g:ale_python_flake8_executable = 'flake8'
let g:ale_python_flake8_options = ''
let g:ale_python_flake8_use_global = 0
let g:ale_python_mypy_executable = 'mypy'
let g:ale_python_mypy_ignore_invalid_syntax = 0
let g:ale_python_mypy_options = ''
let g:ale_python_mypy_use_global = 0
let g:ale_python_pylint_executable = 'pylint'
let g:ale_python_pylint_options = ''
let g:ale_python_pylint_use_global = 0
 Global Variables:
let g:ale_cache_executable_check_failures = 0
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = 50
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 0
let g:ale_lint_on_filetype_changed = 0
let g:ale_lint_on_save = 0
let g:ale_lint_on_text_changed = 'never'
let g:ale_lint_on_insert_leave = 0
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_linters_explicit = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = 0
let g:ale_open_list = 0
let g:ale_pattern_options = {}
let g:ale_pattern_options_enabled = 0
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_type_map = {}
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1

WRITE WHAT WENT WRONG HERE.

So I'm trying to disable ALE linting and I only want it to lint on save. As you can see on the output i have the following options

```
let g:ale_lint_on_enter = 0
let g:ale_lint_on_save = 0
let g:ale_lint_on_text_changed = 'never'
let g:ale_lint_on_filetype_changed = 0
let g:ale_lint_on_insert_leave = 0
```

I'm doing this just to try and disable linting on everything. However none of these options seem  to be respected so I tried replacing all of them with:

```
let g:ale_enabled = 0
```

Which completely disabled ale linting all together so it wasn't an issue of me putting something in the wrong place. I just cannot disable the linting and it runs the linters so often that it severely slows down my computer.

**Comment by w0rp** [comment]
Can you provide some steps for repeating the bug? Did you try to restart vim after putting those settings in your vimrc file?

**Comment by thornycrackers** [comment]
Alright I have tracked it down. I created a brand new Ubuntu 16.04 box. Heres most of the commands I ran to get setup.

```
apt-get update
apt-get install software-properties-common -y
add-apt-repository ppa:neovim-ppa/stable
apt-get update
apt-get install neovim
apt-get install python3-pip -y
pip3 install neovim jedi flake8 flake8-docstrings flake8-isort flake8-quotes pep8-naming pep257 isort mypy ansible-lint
```

I installed plug in the correct directory:
```
wget https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
```

I created an init.vim file with the following:

```
call plug#begin('~/.config/nvim/plugged')
Plug 'w0rp/ale' " Syntax checking
call plug#end()
" Write this in your vimrc file
let g:ale_lint_on_text_changed = 'never'
" You can disable this option too
" if you don't want linters to run on opening a file
let g:ale_lint_on_enter = 0
```

Very simple, installed Ale and this worked correctly as it should. But this isn't where I place my configuration for vim. I actually put it inside a separate vim module. so I changed it to this.

```
call plug#begin('~/.config/nvim/plugged')
Plug 'w0rp/ale' " Syntax checking
Plug 'thornycrackers/vim-options' " Syntax checking
call plug#end()
```

Now if you look at my vim options it's very complicated but after I cloned the repo down I replaced everything in the vim file with the following:

```
" Write this in your vimrc file
let g:ale_lint_on_text_changed = 'never'
" You can disable this option too
" if you don't want linters to run on opening a file
let g:ale_lint_on_enter = 0
```

And then it didnt work 

# TLDR;

The options are respected when they are in `init.vim` but if I moved the options into a separate vim plugin they do not.

**Comment by w0rp** [comment]
That's to be expected. You have to set those options before the plugin loads so ALE can avoid setting up the `autocmd` commands for them, which cuts down on processing time.
